extensions/amp-form/0.1/amp-form.js

```diff
@@ -1223,6 +1229,80 @@ export class AmpForm {
     }
   }
 
+  /**
+   * Initialize form fields from query parameter values if attribute
+   * 'data-initialize-from-url' is present on the form and attribute
+   * 'data-allow-initialization' is present on the field.
+   * @private
+   */
+  maybeInitializeFromUrl_() {
+    if (
+      isProxyOrigin(this.win_.location) ||
+      !this.form_.hasAttribute('data-initialize-from-url')
+    ) {
+      return;
+    }
+
+    const valueTags = ['SELECT', 'TEXTAREA'];
+    const valueInputTypes = [
+      'color',
+      'date',
+      'datetime-local',
+      'email',
+      'hidden',
+      'month',
+      'number',
+      'range',
+      'search',
+      'tel',
+      'text',
+      'time',
+      'url',
+      'week',
+    ];
+    const checkedInputTypes = ['checkbox', 'radio'];
+
+    const queryParams = parseQueryString(this.win_.location.search);
+    Object.keys(queryParams).forEach(key => {
+      const field = this.form_./*OK*/ querySelector(`[name="${key}"]`);
```

Comment:
Can we use `this.form_.elements[key]` here? It would also support non-descendant fields that reference the form by name and might even be a bit faster.

We don't do this consistently in existing amp-form code unfortunately.

Reply:
I like this suggestion, but might need a little help understanding where/why check-types fails.

`form.elements[key]` can return a `NodeList` (for radio inputs) or an `Element`. I took this into account with my refactor of the `maybeInitializeFromUrl_()` function. See [gist](https://gist.github.com/mattwomple/54e956837a84e3c16ce90264761f18a1#file-form-elements-js-L75). Now, the `check-types` test fails:
```
Type checking failed:
extensions/amp-form/0.1/amp-form.js:1297: ERROR - [JSC_TYPE_MISMATCH] actual parameter 1 of iterateCursor$$module$src$dom does not match formal parameter
found   : HTMLElement
required: IArrayLike<?>
missing : [length]
mismatch: []
        iterateCursor(fields, field => maybeFillField(field, key));
```
Does this indicate that the check-types test needs to be updated to understand that `form.elements[key]` can return a `NodeList`? Or is there a way that I can tell the test "hey, i know this is iterable"?

Reply:
Ah, the Closure type for `HTMLFormElement.elements` might be incorrect (missing the list case). Typically we typecast to workaround.

```js
if (fields.length) {
  // Typecast since Closure is missing NodeList union type in HTMLFormElement.elements.
  const list = /** @type {!NodeList} */ (fields);
  [...]
}
```

Reply:
Sounds reasonable, but not allowed:
```
extensions/amp-form/0.1/amp-form.js:1307: ERROR - [JSC_INVALID_CAST] invalid cast - must be a subtype or supertype
from: HTMLElement
to  : NodeList
        const list = /** @type {!NodeList} */ (field);
                                              ^^^^^^^
```

Reply:
Ok, I think I have a handle on this:
```
// Typecast since Closure is missing NodeList union type in HTMLFormElement.elements.
const field = /** @type {(!NodeList|!Element)} */ (this.form_.elements[key]);
```

Reply:
This should be resolved in 1894b76. Thanks much for the review and assistance!